src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs

```diff
@@ -167,13 +156,7 @@ public static LocalizableResourceString GetDescription(ErrorCode code)
 
         public static string GetHelpLink(ErrorCode code)
         {
-            string helpLink;
-            if (s_helpLinksMap.Value.TryGetValue(code, out helpLink))
-            {
-                return helpLink;
-            }
-
-            return string.Empty;
+            return $"https://msdn.microsoft.com/query/dev16.query?appId=Dev16IDEF1&l={CultureInfo.CurrentCulture}&k=k({GetId(code)});k(DevLang-csharp)&rd=true";
```

Comment:
I don't particularly like hardcoding `dev16` into the compiler, I think any link would need to be VS-version agnostic.

What would happen when we add new error codes to the compiler that aren't yet doc'd?

Reply:
> I don't particularly like hardcoding dev16 into the compiler, I think any link would need to be VS-version agnostic.

I don't think it's a big deal. I even don't know how it matters if it's dev16 or dev15 or whatever version. Both redirects to the correct page.

If the docs consolidated all error messages in the same directory, we can rely on it as a different link.

> What would happen when we add new error codes to the compiler that aren't yet doc'd?

It will redirect to https://docs.microsoft.com/en-us/dotnet/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error

(Currently, it won't do that for all error messages. That will get fixed by https://github.com/dotnet/docs/pull/20069

Reply:
i think this needs to go through doc team to know what links they want us to point at.  @jinujoseph @kendrahavens @mikadumont do you know who would be a good contact there to ask about this?

Reply:
> I don't think it's a big deal. I even don't know how it matters if it's dev16 or dev15 or whatever version. Both redirects to the correct page.

I'd prefer not to have _any_ version at all. Errors from the C# compiler are not really version specific.

Reply:
@CyrusNajmabadi The docs team currently relies on f1_keywords metadata that exists in articles. Because the error messages documentation is scattered in two different folders currently, I don't think there is another way.

AFAIK, the docs team has an internal request to support a whole folder redirection, which will make it easy to have all error messages in one folder.

I'll let @BillWagner or @adegeo confirm that.

Reply:
@333fred the version in this link doesn't seem to affect anything. I even don't know why it exists :smile:
What the link do is it redirects to the article which have the appropriate metadata, which is obviously has nothing to do with version.

Reply:
> @333fred the version in this link doesn't seem to affect anything. I even don't know why it exists smile

Then I don't want it in the compiler :slightly_smiling_face:. Version numbers like that are a red flag to me because someone, someday, is going to decide that we don't need to support dev15 or dev16 anymore, and suddenly that will stop working entirely. With very, very few exceptions we do not know anything about specific target frameworks or specific vs versions inside the compiler, and I don't want to change that.

Reply:
Adding @kexugit who can provide F1 help suggestions.

Reply:
@333fred, Then possibly we will have to wait until docs consolidate error messages in one place.
But I see Visual Studio uses that same link when pressing F1 on a keyword. How this is handled inside VS?

Reply:
> How this is handled inside VS?

I honestly have no idea how VS finds its help links today. I just know that that link isn't inside the compiler.

Reply:
A temporary solution would be to hard code the currently-documented error messages with their links. But this can't guarantee that newly-documented messages will be redirected to that link. However, if we return string.Empty for undocumented messages, Visual Studio will use a Bing search.

So, if we used the hard coding solution:

- The currently-documented messages will use the correct link.
- The currently-undocumented messages will use a Bing search, even if they are documented later. (which is the current behavior).

It's certainly not the best solution, but I can't seem to find another solution (unless there is another link that's more reliable - or - if you can submit an internal request for that).

Reply:
I even noticed just know that we can't even rely on docs.microsoft links even after all error messages exist in the same directory, because it will give 404 for undocumented messages, which isn't the case with using f1 keywords.

Reply:
> I honestly have no idea how VS finds its help links today. I just know that that link isn't inside the compiler.

IDE hardcodes a link based on what the doc people told us to use.  effectively: https://github.com/dotnet/roslyn/blob/e69b451b362312c18602beadd5c8b9e93833140d/src/VisualStudio/Core/Def/Implementation/Utilities/BrowserHelper.cs#L48-L65

Reply:
Pinging @kexugit. One question above is whether there is a link format that doesn't require us hard-coding an AppID, or whether we could use a long-lasting AppID for roslyn/compiler.

Reply:
@OsmondJiang Could you confirm if they need to hardcode the AppID or if it doesn't impact the F1 help result with the newly released F1 help service?

Reply:
AppID is hard required for F1 query requests, current the allowed app id list is:

      "Dev14IDEF1",
      "Dev15IDEF1",
      "Dev12IDEF1",
      "Dev11IDEF1",
      "Dev10IDEF1",
      "MsDynAxRetailEssen063",
      "MVC",
      "Dev16IDEF1",
      "Dev17IDEF1",
      "Dev18IDEF1",
      "Dev19IDEF1",
      "Dev20IDEF1"

@kexugit @Youssef1313 I think it's OK to add a new AppId to the list for roslyn/compiler, what do you think?

Reply:
@OsmondJiang I think it would be perfect if it won't have a hard-coded version in it.

@333fred What do you think?

Reply:
What do you mean about the “version”? Is it the dev16 in “dev16.query” or the AppId?

Reply:
@OsmondJiang the "dev16"

Reply:
@Youssef1313 OK, then we can add a new AppId like `Roslyn` for F1 query service, then you can write the query link like that `https://msdn.microsoft.com/query/roslyn.query?appId=Roslyn&l=EN-US&k=k(CS1001);k(DevLang-csharp)&rd=true` What do you think? @kexugit

Reply:
> `https://msdn.microsoft.com/query/roslyn.query?appId=Roslyn&l=EN-US&k=k(CS1001);k(DevLang-csharp)&rd=true`

@333fred Does this URL look good?

Reply:
That url looks good to me.

Reply:
Great!

@OsmondJiang Please ping me when this work is done to update the PR.

Reply:
OK, I will work with @kexugit about the new name and update you soon.

Reply:
We will deploy this change after the Ignite event ends.

Reply:
@Youssef1313 The change has been deployed. You can go ahead and update the PR. Thanks to @OsmondJiang 's team!

Reply:
Thanks a lot @kexugit and @OsmondJiang 🎉 

https://msdn.microsoft.com/query/dev16.query?appId=Dev16IDEF1&l=EN-US&k=k(CS1001);k(DevLang-csharp)&rd=true
https://msdn.microsoft.com/query/roslyn.query?appId=roslyn&l=EN-US&k=k(CS1001);k(DevLang-csharp)&rd=true

At the time of writing this, the first link works, but the second doesn't. Any idea what I'm doing wrong?

Reply:
https://msdn.microsoft.com/query/roslyn.query?appId=roslyn&l=EN-US&k=k(CS1001);k(DevLang-csharp)&rd=true Now it is working.